k8s/kubevirt/kubevirt.go

```diff
@@ -0,0 +1,366 @@
+	}
+	rootDiskName := ""
+	bootDisk, err := c.unstructuredFindKeyValInt64(disks, "bootOrder", 1)
+	if err != nil || bootDisk == nil {
+		return nil, fmt.Errorf("failed to find boot disk in vmi: %w", err)
+	}
+	rootDiskName, found, err = c.unstructuredGetValString(bootDisk, "name")
+	if err != nil || !found || rootDiskName == "" {
+		return nil, fmt.Errorf("failed to find rootDisk name: %w", err)
+	}
+	// Find name of the PVC in VMI. Sample yaml when dataVolume was used by the VMI:
+	// NOTE: the dataVolume may have been garbage collected after pvc was ready.
+	//  spec:
+	//    volumes:
+	//    - dataVolume:
+	//        name: fedora-communist-toucan
+	//      name: rootdisk
+	//
+	volumes, found, err := unstructured.NestedSlice(vmiRaw.Object, "spec", "volumes")
+	if err != nil || !found {
+		return nil, fmt.Errorf("failed to find vmi volumes: %w", err)
+	}
+	pvcName := ""
+	rootVolume, err := c.unstructuredFindKeyValString(volumes, "name", rootDiskName)
+	if err != nil || rootVolume == nil {
+		return nil, fmt.Errorf("failed to find root volume %q in the vmi: %w", rootDiskName, err)
+	}
+
+	// Check if this is a dataVolume or a pvc
+	if dataVolumeRaw, ok := rootVolume["dataVolume"]; ok {
+		dataVolume, ok := dataVolumeRaw.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf(
+				"wrong type for vmi dataVolume: expected map[string]interface{}, actual %T", dataVolumeRaw)
+		}
+		name, found, err := c.unstructuredGetValString(dataVolume, "name")
+		if err != nil || !found {
+			return nil, fmt.Errorf("failed to get name of the rootdisk data volume: %w", err)
+		}
+		// pvc name is always same as the data volume name
+		pvcName = name
+	} else if pvcRaw, ok := rootVolume["persistentVolumeClaim"]; ok {
+		pvc, ok := pvcRaw.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("wrong type for vmi pvc: expected map[string]interface{}, actual %T", pvcRaw)
+		}
+		name, found, err := c.unstructuredGetValString(pvc, "claimName")
+		if err != nil || !found {
+			return nil, fmt.Errorf("failed to get name of the rootdisk pvc: %w", err)
+		}
+		pvcName = name
+	} else {
+		return nil, fmt.Errorf("root volume is neither a dataVolume nor a pvc")
+	}
+	if pvcName == "" {
+		return nil, fmt.Errorf("empty pvc name for the root disk")
+	}
+
+	// check if the VMI is live migratable
+	// Sample yaml:
+	//  status:
+	//    conditions:
+	//    - lastProbeTime: null
+	//      lastTransitionTime: null
+	//      status: "True"
+	//      type: LiveMigratable
+	//
+	liveMigratable := false
+	conditions, found, err := unstructured.NestedSlice(vmiRaw.Object, "status", "conditions")
+	if err != nil {
+		return nil, fmt.Errorf("failed to find conditions in vmi: %w", err)
+	}
+	if found {
+		condition, err := c.unstructuredFindKeyValString(conditions, "type", "LiveMigratable")
+		if err != nil {
+			return nil, fmt.Errorf("failed while finding live migratable condition in vmi: %w", err)
+		}
+		if condition != nil {
```

Comment:
nit.. Maybe we can return err when "condition" is found to be nil. Just like we are returning in case of rootVolume == nil above. I agree that If LiveMigratable condition is not present, we should consider it as false. However we can leave that to the caller to decide and let this function report error. Let me know if you disagree.

Reply:
Discussed offline. We will treat the missing LiveMigratable condition same as LiveMigratable=false for now since it satisfies our current use cases. When we move to the typed client, we can return the original VMI as it is and let the caller decide.